### ISSUE TITLE
How to avoid device->host copy during training when the number of record is in GPU

### ISSUE THREAD

**Issue by rujialiu** [NVlabs/tiny-cuda-nn#81]
First, thanks for the awesome library!

Suppose the training records are generated in another CUDA kernel or DX compute shader, the number of records are saved in GPU memory. However, when calling network->training_step() from host side, we have to know the number of records in order to construct GPUMatrix from raw gpu memory pointer. Of course we can copy this count from device to host, but that introduced a very big latency (for inference, the device->host latency can be 5x slower than inference itself!)

Is there any easy way to avoid device<->host copy completely?

**Comment by Tom94** [comment]
Hi there, making the __tiny-cuda-nn__ batch size depend on a device variable would involve significant changes to the framework, so will likely not happen.

In our applications, we simply pick an upper bound for the number of records and take the performance hit of running __tiny-cuda-nn__ for potentially too many.

For training, in case the number of records is smaller than the bound, we duplicate a few of the generated records to fill up the batch without host synchronization using the `common_device.h:fill_rollover` kernel. (It may pay off to combine it with a permutation based shuffle if the training data distribution is lopsided across the batch.)
